// next-auth/core/routes/signout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = signout;

async function signout(params) {
  const {
    options,
    sessionToken
  } = params;
  const {
    adapter,
    cookies,
    events,
    jwt,
    callbackUrl,
    logger
  } = options;

  if (!sessionToken) {
    return {
      redirect: callbackUrl
    };
  }

  const useJwtSession = options.session.jwt;

  if (useJwtSession) {
    try {
      var _events$signOut;

      const decodedJwt = await jwt.decode({ ...jwt,
        token: sessionToken
      });
      await ((_events$signOut = events.signOut) === null || _events$signOut === void 0 ? void 0 : _events$signOut.call(events, {
        token: decodedJwt
      }));
    } catch (error) {}
  } else {
    try {
      var _events$signOut2;

      const session = await adapter.deleteSession(sessionToken);
      await ((_events$signOut2 = events.signOut) === null || _events$signOut2 === void 0 ? void 0 : _events$signOut2.call(events, {
        session
      }));
    } catch (error) {
      logger.error("SIGNOUT_ERROR", error);
    }
  }

  const sessionCookie = {
    name: cookies.sessionToken.name,
    value: "",
    options: { ...cookies.sessionToken.options,
      maxAge: 0
    }
  };
  return {
    redirect: callbackUrl,
    cookies: [sessionCookie]
  };
}
